Utilities functions mixin

mdo opened the pull request:
Replaces #37114 because I couldn't push to that PR.

> Re-opening of #35977 after accidentally closing it on my repo (sorry again for that). @mdo, I hope I didn't lose any of your edits in the process as I've also rebased `main` to fix conflicts in the docs.
> 
> As a reminder of what it was about:
> 
> > Code for #35945. Just threw some code in to get the shapes of the functions. Still need testing, handling all the cases (values not being maps for example), and docs updates (but I'll do that last once the code is set).
> > The PR adds the following:
> > 
> > 1. functions to get options, a specific option or a specific value for a given utility:
> >    a. `utilities-get-options`
> >    b. `utilities-get-option`
> >    c. `utilities-get-value`
> > 2. mixins to set options (in bulk or a specific one), new values or remove values for a given utility:
> >    a. `utilities-set-options` (defaults to merging with existing, but can be used to completely override by passing `$merge: false`, allowing to completely redefine a utility if necessary)
> >    b. `utilities-set-option` (just a thin wrapper on `utilities-set-options`)
> >    c. `utilities-add-values`
> >    d. `utilities-remove-values`
> > 3. mixins to add or remove utilities (they don't do as much as the rest, but provide consistency)
> >    a. `utilities-add`
> >    b. `utilities-remove`
> > 
> > Please let me know if there'd be one I missed, or if some are too much (2.b, 3.a. and 3.b for example only bring a consistent API by thinly wrapping other existing functions/mixins)

romaricpascal committed:
Add functions and mixins to manipulate the utlities map

- get options, a specific option or a specific value for a given utility
- set options, new values or remove values for a given utility
- add or remove utilities

romaricpascal committed:
Add thorough implementation of utilities functions and mixins

romaricpascal committed:
Update utilities documentation

romaricpascal committed:
Reorganise headings in utilities page and minor edits

mdo committed:
Minor formatting

mdo committed:
Some edits

romaricpascal committed:
Update utility example to avoid cspell error

mdo committed:
Update copy on new utility API docs

mdo commented:
@julien-deramond @louismaximepiton @ffoodd Would love your thoughts on this! Kind of a few PRs worth of history to look through for the full context, but would love to add this for v5.3.

romaricpascal commented:
Happy to answer any question about the code 😄 One of the things buried in the previous PRs (sorry for the mess with my branches) is that the code is checked by [a test suite](https://github.com/romaricpascal/bootstrap/blob/test-utilities-functions-mixins/scss/tests/mixins/_utilities.test.scss?rgh-link-date=2022-03-13T17%3A04%3A09Z) (kept on a separate branch as it'd rely on [sass-true](https://www.npmjs.com/package/sass-true) to have been brought into the project, which is getting sorted in #36029).

louismaximepiton commented on site/content/docs/5.2/customize/color.md:
```diff
@@ -124,26 +124,12 @@ Here's an example that generates text color utilities (e.g., `.text-purple-500`)
-    ),
-  )
-);
+@include utilities-add-values('color', $all-colors);
```
Should use `"` every where right ?

louismaximepiton reviewed (commented):
Just few questions, small things to change IMO and as global comments: 

Should we introduce default values on each function/mixin ?
Should we check each time for existence of the map ?
What's the future for the `maps.scss` file ? I thought it was for this kind of purpose (redefine some maps before using them inside utilities API).

However, this sounds really great to have inside Bootstrap. This doesn't add compiled CSS, might not bring any breaking change, might be really useful when well used ! Great work :)

romaricpascal reviewed (commented):
@louismaximepiton Thanks for the review 😄 I've finally found some time to provide some answers, at least for those related to the internals of the functions/mixins. I can only let you and the Bootstrap team decide on whether functions should be extracted in other files and such 😄 .

I'm not quite sure how to deal with the few amends as it's now @mdo's branch on the Bootstrap repo. A PR towards this branch feels a bit overkill, and it'll likely be much faster if someone with write access moves things around (especially as I don't have as much capacity to contribute nowadays, sorry).

As a side note, now #36029 has been merged, it's likely worth pulling in [the tests I had set up on a separate branch for writing these functions and mixins](https://github.com/romaricpascal/bootstrap/blob/test-utilities-functions-mixins/scss/tests/mixins/_utilities.test.scss).

mdo commented on scss/mixins/_utilities.scss:
```diff
@@ -95,3 +95,127 @@
+//
+
+// Gets the map of options for a given utility
+@function utilities-get-options($utility-name) {
```
Let's leave as-is for now since these functions and mixins are pretty interdependent.

romaricpascal committed:
Add tests for utilities-map functions and mixins

Split the utilities-map helpers in their own file to make tests a little more isolated

romaricpascal committed:
Tidy up utilities map code

- use consistent variable name
- use consistent code structure for functions that do similar things
- remove code duplication for utilities-add

romaricpascal commented:
@louismaximepiton I've tidied up the code in the functions and mixins so the variable names are more consistent and removed code duplication. 

I've also brought in the tests that I had used when writing the code to make sure things stay shipshape. To make the test files not too large, I've split the code for manipulating the utilities map in its own file. I think it's fine for it to have a mix of Sass functions and mixins despite being in the `mixin` folder. It's something that already happens for the breakpoints and border radius and makes it more convenient to import.

romaricpascal committed:
Update documentation for customizing colors

louismaximepiton commented on scss/tests/mixins/_utilities-map.test.scss:
```diff
@@ -0,0 +1,339 @@
+          responsive: true,
+          property: flex,
+          values: (fill: 1 1 auto)
+        ));
```
Wondering if we should use some hardcoded values or something like `map-get($utilities, "flex");`. Since it's done like we could keep it like that for now and change later if needed ?

romaricpascal replied:
It would make the test just rewrite what the function does internally, which I was trying to avoid. It can be changed at a later stage if we feel like it though, like you said 😊

louismaximepiton commented on scss/tests/mixins/_utilities-map.test.scss:
```diff
@@ -0,0 +1,339 @@
+
+@include describe("utilities") {
+
+  @include describe("utilities-get-options") {
```
Do we need to test `map-from` function ?

romaricpascal replied:
I was considering that it could be skipped, given `utility-get-value` was tested thoroughly. Similarly, there's no test for `utility-get-values` as `utility-get-value` runs it under the hood as well. I think the current tests cover all (or at least most to feel safe,  shame there's no way to gather code coverage) of the code in `_utilities-map.scss`. We can always add new tests if we ever refactor and have "utilities-get-values" or "map-from" not called anymore, I think.

louismaximepiton commented on scss/tests/mixins/_utilities-map.test.scss:
```diff
@@ -0,0 +1,339 @@
+      class: margin,
+      values: auto
+    )
+  ) !global;
```
Missing the values inside a map.

romaricpascal replied:
I'm not sure what you mean there, sorry. Can you explain? 😊

louismaximepiton commented on scss/tests/mixins/_utilities-map.test.scss:
```diff
@@ -0,0 +1,339 @@
+      ));
+
+      @include assert-equal(utilities-get-option(flex, responsive), false, "responsive");
+      @include assert-equal(utilities-get-option(flex, rtl), true, "rtl");
```
I'd have checked that the initial values are still here, for instance, `@include assert-equal(utilities-get-option(flex, property), flex, "property");` + Not sure that using another function is a good practice for unit test but I let you decide here !

romaricpascal replied:
Good shout on checking that the function doesn't just wipe out the configuration 🙌🏻 

Regarding not re-using a function, it's not really textbook unit testing to re-use a function to assert, that's true. The function reused is quite light, tested right before so I think it's pretty safe here, especially as it makes the test more readable.

louismaximepiton replied:
Yes OK, I'm fines with that!

louismaximepiton commented on scss/tests/mixins/_utilities-map.test.scss:
```diff
@@ -0,0 +1,339 @@
+      @include assert-equal(utilities-get-options(flex), (
+        class: flex,
+        values: auto none
+      ));
```
Do we need to check that the previous one doesn't exist anymore ?

romaricpascal replied:
The assertion compares that the two map of options are equal so will detect if the previous ones are still there. Had a quick double check to confirm, as it wasn't really clear from the documentations and the following does fail because of the extra value:

```scss
@include it("compares map completely") {
  @include assert-equal((
    value: 1,
    extra-value: 2
  ), (value: 1));
}
```

louismaximepiton replied:
Ok perfect, thanks a lot for checking :smile:

louismaximepiton reviewed (commented):
I think it's way better to read and understand now in the right order (in the scss), thanks for that !
For the API page, I think we should at least keep one example of the previous method to show how that it's still feasible.

Some few questions remains for me, but the work done since last time I reviewed it is insane, well done !

romaricpascal commented on scss/mixins/_utilities.scss:
```diff
@@ -95,3 +95,5 @@
   }
 }
+
+@import "./utilities-map";
```
Happy to move it for consistency with the rest of the files and update the examples in the documentation 😄

romaricpascal commented on site/content/docs/5.3/utilities/api.md:
```diff
@@ -366,25 +366,55 @@ All utilities generated by the API include `!important` to ensure they override
-Now that you're familiar with how the utilities API works, learn how to add your own custom classes and modify our default utilities.
+{{< callout success >}}
+**New in v5.3.x!** We've added new mixins and functions to better help you use our utilities API. These replace our previous guidance of using Sass's built-in `map-merge()` function and add new functionality for [composing components with utilities](#get-values-and-options).
+{{< /callout >}}
```
@louismaximepiton I don't see any special treatment on the "add-to-map" page, so I'm not sure what you mean. I did find a few examples of new features highlighted in 5.2.0:

- with bold text in https://getbootstrap.com/docs/5.2/components/navbar/
- and with a little block in https://getbootstrap.com/docs/5.2/components/badge/#background-colors

Maybe the callout is just the progression for that for when we want not just to say there's a new feature but also add a bit of text to describe what changes? @mdo what was your reason for using a callout?

romaricpascal committed:
Make small amends from PR review

romaricpascal committed:
Refactor to make import of utilities-map explicit in `_mixins.scss`

romaricpascal commented:
Updated again, hopefully I didn't miss anything. Thanks for the comments 😊 

> For the API page, I think we should at least keep one example of the previous method to show how that it's still feasible.

🤔 Ultimately we'd want to encourage people to use those mixins, not tamper with the `$utilities` map directly like before, so I'd lean towards not advertising that they can still `map-merge()`. We could link the "These replace our previous guidance of using Sass's built-in `map-merge()` function" from the callout (if we keep it) to the 5.2 documentation, as a little reminder (which wouldn't hurt in any case I think)

> Some few questions remains for me, but the work done since last time I reviewed it is insane, well done !

Thanks, though that was mostly bringing the tests that were already waiting in a separate branch for the [sass-true PR](https://github.com/twbs/bootstrap/pull/36029) to be merged, so not as big an edit as the diff makes it look like 😆

louismaximepiton reviewed (approved):
Looks fine to me, let's wait for another thought on the way to add some text in a new feature ! Great work :smile:
